xdsl/backend/riscv/lowering/convert_riscv_scf_to_riscv_cf.py

```diff
@@ -0,0 +1,166 @@
+from xdsl.dialects import builtin, riscv, riscv_cf, riscv_scf
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+class LowerRiscvScfForPattern(RewritePattern):
+    """
+    Create a CFG subgraph for the loop around its body blocks (if the body
+    contained other loops, they have been already lowered to a flow of blocks).
+    Maintain the invariants that a CFG subgraph created for any loop has a single
+    entry and a single exit, and that the entry/exit blocks are respectively
+    first/last blocks in the parent region.  The original loop operation is
+    replaced by the initialization operations that set up the initial value of
+    the loop induction variable (%iv) and computes the loop bounds that are loop-
+    invariant for affine loops.  The operations following the original scf.for
+    are split out into a separate continuation (exit) block. A condition block is
+    created before the continuation block. It checks the exit condition of the
+    loop and branches either to the continuation block, or to the first block of
+    the body. The condition block takes as arguments the values of the induction
+    variable followed by loop-carried values. Since it dominates both the body
+    blocks and the continuation block, loop-carried values are visible in all of
+    those blocks. Induction variable modification is appended to the last block
+    of the body (which is the exit block from the body subgraph thanks to the
+    invariant we maintain) along with a branch that loops back to the condition
+    block. Loop-carried values are the loop terminator operands, which are
+    forwarded to the branch.
+
+    ```
+
+         +--------------------------------------------------------------+
+         |   <code before the ForOp>                                    |
```

Comment:
```suggestion                              |
```

I'd remove this as this happens anyway. Same for the exit block below.

Reply:
I think it's a helpful visualisation of the fact that the code before the op is in the same block as the initial %iv value and bge op, similarly for the exit block.

Reply:
OK, no strong feelings about it